### PR TITLE
chore: rm some custom utils impl

### PR DESCRIPTION
**Pull request by longlho**
- Use shallow-equal instead of custom impl
- Convert stateless components to stateless functional components. We already cache Formatters so re-rendering is super cheap.
- remove unused packages
- no need to enforce intl bc withIntl already does that

### CODE REVIEW

**Comment by DragonRaider5** [comment]
Hey @longlho can you please point me to the caching? Thanks in advance!

**Comment by longlho** [comment]
@DragonRaider5 https://github.com/formatjs/react-intl/blob/247dbaef31ce2b9a5118393a4129be602a497b94/src/components/provider.js#L127-L133

File: src/components/html-message.js [hunk 1]
```diff
@@ -9,8 +9,6 @@ import withIntl from './withIntl';
 import {BaseFormattedMessage} from './message';
 
 class FormattedHTMLMessage extends BaseFormattedMessage {
-  static displayName = 'FormattedHTMLMessage';
```
Comment (redonkulus) [inline src/components/html-message.js]: why was this removed?

File: test/unit/components/date.js [hunk 2]
```diff
@@ -67,26 +70,7 @@ describe('<FormattedDate>', () => {
         expect(rendered.type()).toBe('span');
         expect(rendered.text()).toBe(intl.formatDate(date));
     });
-
```
Comment (redonkulus) [inline test/unit/components/date.js]: should we not test this anymore?

Comment (longlho) [reply test/unit/components/date.js]: yeah SFC is always re-rendered but since we cache the `Intl.DateTimeFormat` itself this yields minimal benefits (might be even faster cause u don't have to rehash during `shouldComponentUpdate`)

File: src/components/html-message.js (hunk 1 above)
Comment (longlho) [reply src/components/html-message.js]: explicit `displayName` isn't required for `class`-based components

**Comment by DragonRaider5** [comment]
@longlho this doesn't cache calls to the formatters AFAIK but only the creation of formatters.

**Comment by DragonRaider5** [comment]
See https://github.com/formatjs/formatjs/blob/d3869feb0018f87eab32280cdc64fb086c964a26/packages/intl-format-cache/src/index.ts#L68

**Comment by DragonRaider5** [comment]
@longlho @redonkulus please roll back these changes until this is clarified.

Or, as @redonkulus approved, maybe he can tell me where exactly the formatter calls are being cached.
The creation is, but the formatters aren't (and it would be stupid to cache those).

File: test/unit/components/date.js (hunk 2 above)
Comment (DragonRaider5) [reply test/unit/components/date.js]: The test and the SCU are both required, as this is not being cached.

**Comment by longlho** [comment]
Why do you need to cache format calls? Those are not the bottlenecks
(cached constructor can calls 1M format per sec)

**Comment by longlho** [comment]
https://github.com/formatjs/formatjs/tree/master/packages/intl-format-cache#benchmark

**Comment by DragonRaider5** [comment]
Yes but those are the **constructors**. It's not the actual formatting calls!

**Comment by DragonRaider5** [comment]
You can also see so in the benchmark suite. Only the creation of formatters is being benched, not the formatting itself (which is what we try to minimize using SCU).

**Comment by longlho** [comment]
Yeah sry that was missing but I added the benchmarks for formatting itself (including cache lookup):
```
number x 536,024 ops/sec ±0.99% (86 runs sampled)
datetime x 397,275 ops/sec ±0.92% (90 runs sampled)
messageformat x 1,278,072 ops/sec ±1.31% (89 runs sampled)
--- all formats: Fastest is messageformat ---

number x 532,863 ops/sec ±0.79% (93 runs sampled)
datetime x 377,391 ops/sec ±1.11% (89 runs sampled)
messageformat x 709,020 ops/sec ±3.19% (81 runs sampled)
--- all formats random input: Fastest is messageformat ---
```

**Comment by DragonRaider5** [comment]
Those benchmarks aren't representative, as they change with the size of the translations object.
That's why I think SCU is needed.

To get actual values, we'd have to sample the formatting functions with variable message counts and reasonable message lengths (+ variables) and extrapolate the obtained data in order to make an estimation for real world performance. 

I think we should just stick to the SCUs, rest of the PR is great :tada::heart:

**Comment by longlho** [comment]
FormattedMessage still has sCU btw :) that's not gone

**Comment by redonkulus** [comment]
@DragonRaider5 I've published this change in `v3.0.0-beta-8` to allow users a chance to test this in their apps. If this causes performance degradation, we can revert this change.
